tracer/src/Datadog.Trace/Iast/IastModule.Escape.cs

```diff
@@ -0,0 +1,59 @@
+// <copyright file="IastModule.Escape.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Net;
+using Datadog.Trace.Configuration;
+
+namespace Datadog.Trace.Iast;
+
+internal static partial class IastModule
+{
+    public static string? OnXssEscape(string? text)
+    {
+        var res = WebUtility.HtmlEncode(text);
+        try
+        {
+            if (!iastSettings.Enabled || string.IsNullOrEmpty(text))
+            {
+                return res;
+            }
+
+            var tracer = Tracer.Instance;
+            if (!tracer.Settings.IsIntegrationEnabled(IntegrationId.Xss))
+            {
+                return res;
+            }
+
+            var scope = tracer.ActiveScope as Scope;
+            var traceContext = scope?.Span?.Context?.TraceContext;
+
+            if (traceContext?.IastRequestContext?.AddVulnerabilitiesAllowed() != true)
+            {
+                return res;
+            }
+
+            var tainted = traceContext?.IastRequestContext?.GetTainted(text!);
+            if (tainted is null)
+            {
+                return res;
+            }
+
+            if (object.ReferenceEquals(res, text))
+            {
+                // If returned reference is the same, we create a new one
+                res = new string(text!.ToCharArray());
```

Comment:
NIT: Might be worth optimizing this to avoid the array allocation (can be a separate PR).

On .NET 6 there is a constructor that takes a `ReadOnlySpan<char>`.

For other versions of .NET you can use unsafe code:

```
        private static unsafe string CopyString(string source)
        {
            fixed (char* c = source)
            {
                return new string(c, 0, source.Length);
            }
        }
```